### ISSUE TITLE
Allow configuration of `cleanUrls` option for `serve` command

### ISSUE THREAD

**Issue by SlicedSilver** [facebook/docusaurus#7991]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Description

Allow configuration of the `cleanUrls` option passed to [`serve-handler`](https://github.com/vercel/serve-handler) package when running `docusaurus serve`.

Currently, Docusaurus sets the `cleanUrls` option for the `serve-hander` package to `true` and it isn't configurable by the user. 
https://github.com/facebook/docusaurus/blob/bd9f29a4698963d3b139f7419877a300640b09ff/packages/docusaurus/src/commands/serve.ts#L76-L82

Proposing that this should be a user-configurable option.

### Motivation

By default, at least a few major hosting providers (GH Pages, Netlify, Vercel) don't enable this 'feature'. Having this as an option would allow for more accurate testing of the documentation website before deploying.

**A Use Case:**
It is possible to embed an iframe into a documentation page which serves as a self-contained example. The example file can be loaded using the `file-loader` web pack loader. 
```mdx
<iframe
     src={require('!!file-loader!./assets/example.html').default}
 ></iframe>
 <a href={require('!!file-loader!./assets/example.html').default} target="\_blank">
     View example in a new window
 </a>
```

This works correctly when using `docusaurus start` and when deploying the build onto a hosting provider. However, when running `docusaurus serve`, any url ending with `.html` is rewritten without the extension and a 404 error occurs.

### API design

`ServeCLIOptions` could include `cleanUrls?: boolean;` which defaults to `true`.
https://github.com/facebook/docusaurus/blob/bd9f29a4698963d3b139f7419877a300640b09ff/packages/docusaurus/src/commands/serve.ts#L20-L25

When the user wishes to disable `cleanUrls` then they would call the serve command as follows:
`docusaurus serve --no-clean-urls`

A similar options could be added to the configuration file (https://github.com/facebook/docusaurus/blob/bd9f29a4698963d3b139f7419877a300640b09ff/packages/docusaurus-types/src/config.d.ts).

### Have you tried building it?

I've tried the changes locally and passing `cleanUrls: false` to `serve-handler` works as expected.

### Self-service

- [X] I'd be willing to contribute this feature to Docusaurus myself.

**Comment by Josh-Cena** [comment]
Maybe we could let you pass a JSON string through command line? I don't have good API designs so far.

**Comment by SlicedSilver** [comment]
If the goal is to make the entire `serve-handler` options object customisable from the command line then it may be a bit cumbersome to achieve that with a JSON string.

For example:
`docusaurus serve --serve-options='{"cleanUrls":false, "renderSingle": false, "directoryListing": ["/assets/**", "/!assets/private"]}'`

Especially if trying to customise the `headers` option (https://github.com/vercel/serve-handler#headers-array) which can be rather lengthy.

### Alternatives

* Might be better to allow the loading of a JSON file from disk, much like the `config` cli option for the other commands:
`docusaurus serve --serve-config ./serve-config.json`

* or include these options within the main `docusaurus.config.js` file.

I'm happy to work on a PR if there is a decision on the preferred approach for the API.

**Comment by Josh-Cena** [comment]
Yeah, I think a `serveOptions` in the config file makes more sense, just like we have several options for `deploy` as well. I'm not totally decided yet.

**Comment by slorber** [comment]
To me this CLI command is just a convenience for non-technical users to easily test their Docusaurus site. If you don't like it or need something more advanced you can provide your own server.

What prevents you from running `npx serve` and have a `serve.json` config file for example? Docusaurus doesn't do much more than that.

**Comment by SlicedSilver** [comment]
Agreed, that makes sense. It would be out of scope to add all those options from a third-party lib.

I can use `npx serve` or create my own script to achieve what I'm looking for. 

Would it make sense to change the default from `true` to `false` which will match the behaviour of the `start` command and the majority of hosting providers which I've checked?

Alternatively, could the script check for an environmental variable `CLEAN_URLS=false` ? No need to change the API then.

Happy that this issue can be closed if deemed out of scope.👍

**Comment by slorber** [comment]
> Would it make sense to change the default from true to false which will match the behaviour of the start command and the majority of hosting providers which I've checked?

As far as I know, Netlify has pretty URLs by default, although it tells you it doesn't (asset optimization disabled by default, but still pretty URLs enabled)

I've documented this here:
- https://github.com/slorber/trailing-slash-guide
- https://github.com/slorber/trailing-slash-guide/blob/main/docs/Hosting-Providers.md#netlify

This is confusing but it's what I see in practice.

Also, have you tried deploying a Docusaurus site without pretty URLs? Would our frontend SPA router work with URLs ending with `xyz/index.html` or `xyz.html` for example? 

If you want to change this setting, I'd rather ask you to do the homework and show me a Docusaurus site hosted on multiple platforms with cleanUrls: false and trailingSlash true/false combinations.

---

> Alternatively, could the script check for an environmental variable CLEAN_URLS=false ? No need to change the API then.

That's not particularly better to me. If this is a useful option we should expose it properly. But is it a useful option in the first place?

--- 

I'm still not sure to understand your iframe use-case. What is the easy actionable step-by-step guide for reproducing this problem?

**Comment by Josh-Cena** [comment]
I don't think it's quite out of scope. If some hosting providers have this behavior and others don't, we ought to provide an option to make the local testing env mirror the deployment env.

**Comment by slorber** [comment]
> I don't think it's quite out of scope. If some hosting providers have this behavior and others don't, we ought to provide an option to make the local testing env mirror the deployment env.

I'm ok to add an option for this

But if some hosting provider decides to do weird fancy things, I wouldn't make it part of `docusaurus serve` scope. For example GitHub Pages without a `.nojekyll` file would have a quite surprising static serving behavior, and we definitively don't want to add support for that

**Comment by Josh-Cena** [comment]
Aye, yes, that makes sense

**Comment by SlicedSilver** [comment]
I've created a stackbiltz example where the site functions correctly on `npm run start`, but not when running `npm run serve`. 

https://stackblitz.com/edit/github-sifwoz?file=docusaurus.config.js

**It appears that this behaviour only occurs when `baseUrl` is changed from the default of `'/'`.**

---

The example:
- Click 'View Example' on the landing page of the documentation site.
- When run with `start` then you will see a light blue iframe
- When run with `serve` then you will see the landing page within the iframe.

**Comment by rashidmya** [comment]
can I try doing this?

**Comment by slorber** [comment]
@SlicedSilver is it possible that it's related to our `serve.ts` code having some "extras"? 

![CleanShot 2022-08-26 at 17 46 16@2x](https://user-images.githubusercontent.com/749374/186943564-46d1e6d9-db32-4aad-8044-64deff72289d.png)

--- 

@rashidmya yes I'm ok to add a `--no-clean-urls` option. 
But let's wait first to be sure that it's needed, cf my comment above. You can help investigate the problem

**Comment by SlicedSilver** [comment]
Within `serve.ts`, if I set `cleanUrls: false` then it works as expected.

However, if I remove the http 302 'found' redirect logic then it works as well (without changing the `cleanUrls`):
```js
    // if (!req.url?.startsWith(baseUrl)) {
    //   res.writeHead(302, {
    //    Location: baseUrl,
    //  });
    //  res.end();
    //  return;
    // }
```
Changing the `Location` value within the redirect also works (without changing the `cleanUrls`):
```js
Location: baseUrl + req.url.slice(1),
```

If changing the location value is the best option then obviously it might make sense to check if baseUrl ends with `/` and url starts with `/` if that is quietly allowed within the configuration.
```js
const slicePoint = (baseUrl.endsWith('/') && req.url.startsWith('/')) ? 1 : 0;
// ...
Location: baseUrl + req.url.slice(slicePoint),
```

@slorber Hope this helps, and thanks for your assistance.

**Comment by MattiasBuelens** [comment]
This tripped us up too.

We have some `.html` files inside `/static/examples/`, and we put them in an `<iframe>` inside our docs:
```mdx
<iframe src={useBaseUrl("/examples/demo.html")}></iframe>
```
This works fine with `docusaurus start`, but breaks with `docusaurus build && docusaurus serve`. We found that changing `baseUrl` from `/docs/` to `/` fixes the issue, but unfortunately we can't do that for this website.

Reproduction case: https://github.com/MattiasBuelens/docusaurus-serve-html-asset-issue

> What prevents you from running `npx serve` and have a [`serve.json`](https://github.com/vercel/serve#configuration) config file for example? Docusaurus doesn't do much more than that.

That seems like a reasonable workaround for us. I'd prefer if `docusaurus serve` would have better support for `baseUrl`, but for now we'll use a custom `serve` script instead.

**Comment by afrinashar** [comment]
I would like to work on this.

**Comment by slorber** [comment]
@afrinashar we already have a PR: https://github.com/facebook/docusaurus/pull/10552

**Comment by slorber** [comment]
I don't think we can proceed and implement this feature properly.

@HD787 implemented it here in the CLI: https://github.com/facebook/docusaurus/pull/10842

The problem is that our website uses a router with paths such as `/docs/xyz`

Using `--no-clean-urls` would mean that you disable clean URLs for all your host.

This means your Docusaurus pages must be accessed using the `doc.html` extension.

`http://localhost:3000/index.html` works, but it's ugly, and navigating to any page will remove the extension so you break the deep linking/refresh pattern.

`http://localhost:3000/` doesn't really work anymore and leads to serving a 404 page initially, which gets fixed once React hydrates:

https://github.com/user-attachments/assets/2ab301db-7598-4db2-b47e-3ae86306d698

---

`docusaurus serve` is only here to conveniently emulate a Docusaurus site in production mode. 

It won't be able to cover all the use cases where you embed other static content inside an existing site, in particular when this other content needs to obey different hosting rules that our Docusaurus static files.

Note that Docusaurus is just a tiny wrapper around [serve-handler](https://github.com/vercel/serve-handler), and for more advanced needs you'd rather create your own custom little server.

`docusaurus serve` is mostly just this:

```js
const handler = require('serve-handler');
const http = require('http');

const server = http.createServer((request, response) => {
  serveHandler(req, res, {
    public: "build",
    trailingSlash: require("docusaurus.config.js").trailingSlash,
    directoryListing: false,
  });
});

server.listen(3000, () => {
  console.log('Running at http://localhost:3000');
});
```

For advanced needs, you can use different "serve handlers" depending on the request path, each using a different configuration.

---

Or use the [serve](https://github.com/vercel/serve) CLI and provide a `serve.json` config: `npx serve build`

Eventually replace our `docusaurus serve` command with another one you can customize:

```diff
-  "serve": "docusaurus serve"
+  "serve": "serve build"
```

This allows you to configure your own hosting rules, rewrites and redirects.

---

I don't think it's reasonable for Docusaurus to provide advanced `docusaurus serve` config that would cover all needs. CLI arguments may not be the ideal way to provide that config either. In advanced cases, you'd likely need to combine multiple serve handlers with different configurations.

For this reason I'm going to close and consider this has to be implemented in userland. `docusaurus serve` is a best effort solution to solve our own site hosting problem, but we can't cover "mixed deployments" that must obey hosting rules that diverge.

**Comment by bmaupin** [comment]
I figured out this workaround using [http-server](https://github.com/http-party/http-server):

1. Install http-server from a specific commit ([it's not had a new release since `--base-dir` was added](https://github.com/http-party/http-server/issues/947))
   ```
   npm i --save-dev "https://github.com/http-party/http-server.git#741ae2b"
   ```
1. Modify the `serve` command in package.json
   ```
   "serve": "http-server build --base-dir BASEDIR",
   ```
   - Replace `BASEDIR` with your base URL without the slashes (e.g. `docs`)

Or you can run it without modifying the serve script:

```
npx https://github.com/http-party/http-server.git#741ae2b build --base-dir BASEDIR
```

I wouldn't recommend this for any kind of production usage of course.
